CASSANDRA-16362 Correct SSLContext initialization order for sstableloader

jonmeredith opened the pull request.

jonmeredith committed:
Correct initialization order for createSSLContext

The SSLContext needs to be initialized before the default parameters
can be adjusted. (Bug introduced in CASSANDRA-13325).

Patch by Jon Meredith; Reviewed by David Capwell for CASSANDRA-16362

jonmeredith committed:
DNM: Setting CircleCI to high res config

jonmeredith committed:
Add test, override JDKSSLOptions until accepts protocols can be added

Setting the default SSLContext default options has no effect

dcapwell reviewed (commented):
this isn't enough.  I finally got a test working and we need the following as well.

```
diff --git a/src/java/org/apache/cassandra/tools/BulkLoader.java b/src/java/org/apache/cassandra/tools/BulkLoader.java
index 5fd3885465..b3dfd59447 100644
--- a/src/java/org/apache/cassandra/tools/BulkLoader.java
+++ b/src/java/org/apache/cassandra/tools/BulkLoader.java
@@ -19,6 +19,7 @@ package org.apache.cassandra.tools;

 import java.io.IOException;
 import java.net.InetSocketAddress;
+import java.util.Arrays;
 import java.util.Set;
 import javax.net.ssl.SSLContext;

@@ -261,10 +262,11 @@ public class BulkLoader
             throw new RuntimeException("Could not create SSL Context.", e);
         }

-        return JdkSSLOptions.builder()
-                            .withSSLContext(sslContext)
-                            .withCipherSuites(clientEncryptionOptions.cipherSuitesArray())
-                            .build();
+        JdkSSLOptions.Builder builder = JdkSSLOptions.builder().withSSLContext(sslContext);
+        String[] cipherSuites = clientEncryptionOptions.cipherSuitesArray();
+        if (cipherSuites != null && cipherSuites.length > 0)
+            builder = builder.withCipherSuites(cipherSuites);
+        return builder.build();
     }

     static class ExternalClient extends NativeSSTableLoaderClient
```


Here is my test case

```
@Test
    public void connectionCannotAgreeOnClientAndServerTest() throws Throwable
    {
        try (Cluster cluster = builder().withNodes(1).withConfig(c -> {
            c.with(Feature.NATIVE_PROTOCOL);
            c.set("client_encryption_options",
                  ImmutableMap.builder().putAll(validKeystore)
                              .put("enabled", true)
                              .put("accepted_protocols", Collections.singletonList("TLSv1.2"))
                              .put("cipher_suites", Collections.singletonList("TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384"))
                              .build());
        }).start())
        {
            InetAddress address = cluster.get(1).config().broadcastAddress().getAddress();
            int port = (int) cluster.get(1).config().get("native_transport_port");

            TlsConnection connection = new TlsConnection(address.getHostAddress(), port,
                                                         Collections.singletonList("TLSv1.2"),
                                                         Collections.singletonList("TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"));
            Assert.assertEquals("Should not be possible to establish a TLSv1.2 connection with different ciphers",
                                ConnectResult.FAILED_TO_NEGOTIATE, connection.connect());
            connection.assertReceivedHandshakeException();
        }
    }
```

jonmeredith committed:
switched to finding exception message instead due to cross-platform issues

jonmeredith committed:
filter accepted protocols by supported protocols in bulk loader

jonmeredith committed:
abandoned filtering and went for looking up what the TLS protocols would be

jonmeredith committed:
fix SettingsTableTest to use tlsInstanceProtocolSubstitution

jonmeredith committed:
apply bulkloader accepted protocol fix to stress

jonmeredith commented:
I've reworked the patch to remove the hard-coded list of protocols when the legacy "TLS" protocol is selected. Instead the protocol list is created from `SSLContext.getInstance("TLS")` directly during initialization.
This does have the consequence of including SSLFactory as an import for `DatabaseDescriptor`, but I don't think that causes any significant issues.
The stress tool also has the same issues as BulkLoader so I've also updated that to apply protocols correctly.

dcapwell commented on src/java/org/apache/cassandra/tools/BulkLoader.java:
```diff
@@ -261,10 +268,22 @@ private static SSLOptions buildSSLOptions(EncryptionOptions clientEncryptionOpti
-                            .withSSLContext(sslContext)
-                            .withCipherSuites(clientEncryptionOptions.cipherSuitesArray())
-                            .build();
+        // Temporarily override newSSLEngine to set accepted protocols until it is added to JdkSSLOptions
```
`s/JdkSSLOptions/RemoteEndpointAwareJdkSSLOptions/g`

dcapwell commented on test/unit/org/apache/cassandra/db/virtual/SettingsTableTest.java:
```diff
@@ -154,7 +155,7 @@ public void testEncryptionOverride() throws Throwable
 
         config.server_encryption_options = config.server_encryption_options.withProtocol("TLS");
-        check(pre + "protocol", "[TLSv1.3, TLSv1.2, TLSv1.1, TLSv1]");
+        check(pre + "protocol", "[" + String.join(", ", SSLFactory.tlsInstanceProtocolSubstitution()) + "]");
```
isn't this simpler?

```
check(pre + "protocol", SSLFactory.tlsInstanceProtocolSubstitution()).toString());
```

jonmeredith replied:
Agreed, didn't think to try the default formatting.

dcapwell commented on test/distributed/org/apache/cassandra/distributed/test/SSTableLoaderEncryptionOptionsTest.java:
```diff
@@ -0,0 +1,105 @@
+
+public class SSTableLoaderEncryptionOptionsTest extends AbstractEncryptionOptionsImpl
+{
+    static Cluster cluster;
```
nit: statics should be all caps

dcapwell commented on test/distributed/org/apache/cassandra/distributed/test/SSTableLoaderEncryptionOptionsTest.java:
```diff
@@ -0,0 +1,105 @@
+    @AfterClass
+    public static void tearDownCluster()
+    {
+        cluster.close();
```
this could be null, so should add a null check (if setup fails this is null)

jonmeredith replied:
ack

jonmeredith committed:
review nits from dcapwell

jonmeredith commented:
Merged.
